### PR TITLE
rewrite wysiwyg property updating

### PR DESCRIPTION
**Pull request by dwelle**
So, at first I started making a few fixes which then turned into a full rewrite :smile:.

First, I wanted to fix problems with previous solution that discarded wysiwyg's `blur` event altogether. Problem with finalizing wysiwyg on explicit `pointerDown` on canvas is that we don't finalize when e.g. changing tools (which we should), or zooming in/out (which resulted in visual bugs) and who knows what else.

So instead I used a different workaround: when clicking inside the menu, temporarily remove the `blur` event, and re-enable it on `pointerup` (plus some logic on top to handle edge-cases).

I've also fixed a case when you click on the same property which doesn't result in update, so on previous solution the wysiwyg wasn't focused.

Another bug this PR fixes is trying to type into color inputs, which previously resulted in wysiwyg stealing focus again.

And on top, I enabled this only for desktop shape menu, because previously it resulted in some very weird stuff going on when you tried to edit wysiwyg props on mobile (there I'd argue it should blur & confirm on trying to open mobile menu).

Second, I wanted to simplify the wysiwyg updating. I realized we have a global element mutation event we can subscribe to, so I used that. For now, I'm updating the wysiwyg on every mutation update, which is a bit inefficient but it doesn't matter for now.

Both of the above changes resulted in not having to expose any wysiwyg API outside, and not needing any extra `appState` properties either.

This way all logic related to `wysiwyg` is encapsulated in the module, which I think is a good thing.

Now, I'm not sure if this is 100% working solution, so try to break it and let me know.

### CODE REVIEW

**Comment by kbariotis** [comment]
One thing I found is that color makes the text element to disappear like its color becomes none.

**Comment by kbariotis** [comment]
Tbh, I'm not sure how that compares to the previous approach, it seems a bit more computational heavy since it needs to iterate all elements from the scene to find which one to take styles from and also has to do all these bind/rebind/unbind. 

I understand the approach of keeping the main state as the single source of truth though which can make things easier.

It would be interesting I think if we could measure these kind of things and compare with different approaches. Just an idea.

In any case, it looks good, I only played with it for a few minutes so let us give it another look tomorrow. 👍 great work! :)

**Comment by dwelle** [comment]
It's not that heavy, each loop takes between `0.1-0.8ms` and happens only once per keystroke, and twice per prop update. It can be optimized later, if need be.

The main point of this PR is to fix several bugs, some of which would be hard to fix with previous solution.

It also encapsulates all logic within the module.

As for the color picker bug.. mhm, lemme see.

**Comment by kbariotis** [comment]
Well it’s O(n) compared to the previous O(1). 

One question, what do you mean the loop happen on each keystroke? Is that on every type inside the text element? Can’t we do that on every time a property changes?

**Comment by dwelle** [comment]
> Well it’s O(n) compared to the previous O(1).

Yea, but as Pete Hunt pointed out, it's negligible. E.g., we do this for every `globalSceneState.getElements()`, which is a lot (look it up in the codebase).

And I *am* looking into ways how to optimize it.

> One question, what do you mean the loop happen on each keystroke? Is that on every type inside the text element? Can’t we do that on every time a property changes?

I initially wanted to do that, but that would be non-deterministic because e.g. most props changes trigger 2 updates (we care about the 2nd), but some may do even more. And we can distinguish between prop change updates we care in this wysiwyg case.

**Comment by dwelle** [comment]
Of course I can pass `appState` into the wysiwyg, and use `state.editingElement`. I just didn't bother because it doesn't matter right now → it'd be premature optim. Wysiwyg isn't a tight loop we need to optimize.

**Comment by vjeux** [comment]
Whatever you did, it seems to be fixing the issue when zooming in / out on mobile. I can’t repro on this branch but can super quickly on master

ACE3E28E-7D02-496D-B93F-9758C25BBB39

File: public/index.html [hunk 1]
```diff
@@ -62,10 +62,10 @@
     />
     <!-- OG tags require absolute url for images -->
     <meta name="twitter:image" content="https://excalidraw.com/og-image.png" />
-    <meta
+    <!-- <meta
```
Comment (vjeux) [inline public/index.html:65]: Need to revert before committing

Comment (dwelle) [reply public/index.html:65]: oh.. I disabled in dev, and committed by mistake :)

**Comment by dwelle** [comment]
Ad the DOM removal. Not sure what I did, but I just pushed a safeguard on top to ensure we don't run `cleanup` twice.

Ready for merge.

File: src/element/textWysiwyg.tsx [hunk 2]
```diff
@@ -136,25 +139,74 @@ export function textWysiwyg({
...
+      editable.focus();
+    });
+  };
+
+  // prevent blur when changing properties from the menu
+  const onPointerDown = (event: MouseEvent) => {
+    if (
+      event.target instanceof HTMLElement &&
+      event.target.closest(".App-menu__left") &&
+      !isWritableElement(event.target)
```
Comment (kbariotis) [inline src/element/textWysiwyg.tsx:178]: A more elegant solution maybe here? We shouldn't treat these class names as permanent identifiers.

File: src/element/textWysiwyg.tsx [hunk 3]
```diff
@@ -136,25 +139,74 @@ export function textWysiwyg({
...
+  const unbindUpdate = globalSceneState.addCallback(() => {
+    const editingElement = globalSceneState
+      .getElementsIncludingDeleted()
+      .find((element) => element.id === id);
+    if (editingElement && isTextElement(editingElement)) {
+      Object.assign(editable.style, {
+        font: editingElement.font,
+        textAlign: editingElement.textAlign,
+        color: editingElement.strokeColor,
+        opacity: editingElement.opacity,
```
Comment (kbariotis) [inline src/element/textWysiwyg.tsx:198]: Opacity is not working and I'm not sure why either. I wonder if is being ignored by the contenteditable? Not sure.

Comment (dwelle) [reply src/element/textWysiwyg.tsx:198]: My bad. Fixed.

File: src/element/textWysiwyg.tsx (hunk 2 above)
Comment (dwelle) [reply src/element/textWysiwyg.tsx:178]: I'll maybe add the class to `constants.js`?

Comment (dwelle) [reply src/element/textWysiwyg.tsx:178]: And to guard against regressions, we'll need to test this, which is currently blocked by react-testing-library update.

Comment (dwelle) [reply src/element/textWysiwyg.tsx:178]: Actually, we should already have a version of `jsdom` that supports selection, but it doesn't work. `npm ls jsdom` returns some weird things so don't know.

**Comment by dwelle** [comment]
@kbariotis @lipis https://github.com/excalidraw/excalidraw/pull/1387/commits/f112ebc72d75a80ca9a37192101ca0951c17cb0a

**Comment by lipis** [comment]
@dwelle No opinion here, in general I prefer CSS in JS.. :) No classes. :)

**Comment by dwelle** [comment]
@lipis well the class was there to begin with, I just used it for the JS logic (which also has nothing to do with styling).

**Pull request merged**